File: packages/harness/src/reviewer.ts
/**
 * @create-something/harness
 *
 * Reviewer: Runs peer review sessions using Claude Code.
 * Philosophy: Specialized reviewers analyze code for specific concerns.
 */

import { spawn } from 'node:child_process';
import { writeFile, unlink, readFile } from 'node:fs/promises';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import type {
  ReviewerConfig,
  ReviewerDefinition,
  ReviewContext,
  ReviewResult,
  ReviewFinding,
  ReviewOutcome,
  FindingSeverity,
  ReviewerType,
} from './types.js';
import { getPromptForReviewer } from './reviewer-prompts.js';
import {
  getEffectiveModel,
  type FailureTracker,
} from './failure-handler.js';

// ─────────────────────────────────────────────────────────────────────────────
// Model Selection
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Get the appropriate model for a reviewer type.
 *
 * Philosophy: Different review types have different cognitive demands.
 * - Security: Pattern detection (vulnerabilities, secrets) → Haiku (fast, cheap)
 * - Architecture: Deep analysis (DRY, coupling, design) → Opus (thorough, expensive)
 * - Quality: Balanced review (conventions, tests) → Sonnet (middle ground)
 * - Custom: User-defined, default to Sonnet
 *
 * @param type - The reviewer type
 * @param override - Optional explicit model override from config
 */
export function getReviewerModel(
  type: ReviewerType,
  override?: 'haiku' | 'sonnet' | 'opus'
): 'haiku' | 'sonnet' | 'opus' {
  // Explicit override from config takes precedence
  if (override) {
    return override;
  }

  // Default routing based on reviewer type
  switch (type) {
    case 'security':
      return 'haiku'; // Pattern scanning for known vulnerabilities
    case 'architecture':
      return 'opus'; // Complex reasoning about system design
    case 'quality':
      return 'sonnet'; // Balanced review of conventions and tests
    case 'custom':
      return 'sonnet'; // Safe default for user-defined reviewers
    default:
      return 'sonnet';
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Config Conversion
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Convert a ReviewerDefinition from HarnessConfig to ReviewerConfig.
 * The `prompt` field from config (which may have been loaded from an external file)
 * becomes the `customPrompt` that overrides the default prompt.
 */
export function convertDefinitionToConfig(def: ReviewerDefinition): ReviewerConfig {
  return {
    id: def.id,
    type: def.type,
    enabled: def.enabled,
    canBlock: def.canBlock,
    customPrompt: def.prompt, // Loaded external prompt becomes customPrompt
    includePatterns: def.includePatterns,
    excludePatterns: def.excludePatterns,
    model: def.model, // Optional model override
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// Prompt Generation
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Generate the full prompt for a reviewer session.
 */
export function generateReviewPrompt(
  config: ReviewerConfig,
  context: ReviewContext
): string {
  const basePrompt = getPromptForReviewer(config.type, config.customPrompt);

  // Build context section
  const contextLines: string[] = [];

  contextLines.push('## Files Changed');
  if (context.filesChanged.length > 0) {
    for (const file of context.filesChanged) {
      contextLines.push(`- ${file}`);
    }
  } else {
    contextLines.push('(No files changed)');
  }
  contextLines.push('');

  contextLines.push('## Git Diff (This Checkpoint)');
  contextLines.push('```diff');
  // Truncate diff to avoid context overflow (30k chars for checkpoint)
  const truncatedDiff = context.gitDiff.slice(0, 30000);
  contextLines.push(truncatedDiff);
  if (context.gitDiff.length > 30000) {
    contextLines.push('... (diff truncated)');
  }
  contextLines.push('```');
  contextLines.push('');

  // For architecture reviewer, include full harness diff for DRY detection
  if (config.type === 'architecture' && context.fullHarnessDiff) {
    contextLines.push('## Full Harness Diff (All Changes Since Branch Start)');
    contextLines.push('**IMPORTANT: Check this for DRY violations across all files modified in this harness run.**');
    contextLines.push('```diff');
    // Allocate 20k chars for full harness diff
    const truncatedFullDiff = context.fullHarnessDiff.slice(0, 20000);
    contextLines.push(truncatedFullDiff);
    if (context.fullHarnessDiff.length > 20000) {
      contextLines.push('... (full harness diff truncated)');
    }
    contextLines.push('```');
    contextLines.push('');
  }

  contextLines.push('## Completed Issues');
  if (context.completedIssues.length > 0) {
    for (const issue of context.completedIssues) {
      contextLines.push(`- ${issue.id}: ${issue.title}`);
    }
  } else {
    contextLines.push('(No issues completed)');
  }
  contextLines.push('');

  contextLines.push('## Checkpoint Summary');
  contextLines.push(context.checkpointSummary || '(No summary available)');

  const contextSection = contextLines.join('\n');

  return basePrompt.replace('{CONTEXT}', contextSection);
}

// ─────────────────────────────────────────────────────────────────────────────
// Reviewer Execution
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Run a single reviewer session.
 */
export async function runReviewer(
  config: ReviewerConfig,
  context: ReviewContext,
  options: { cwd: string; timeoutMs: number; failureTracker?: FailureTracker }
): Promise<ReviewResult> {
  const startTime = Date.now();

  const prompt = generateReviewPrompt(config, context);
  const promptFile = join(tmpdir(), `review-${config.id}-${Date.now()}.md`);

  try {
    await writeFile(promptFile, prompt, 'utf-8');

    // Determine effective model (with escalation if failures exist)
    const heuristicModel = getReviewerModel(config.type, config.model);
    let model = heuristicModel;
    let escalated = false;

    if (options.failureTracker) {
      const escalation = getEffectiveModel(
        options.failureTracker,
        config.id, // Use reviewer ID as issue ID for tracking
        heuristicModel
      );
      model = escalation.model;
      escalated = escalation.escalated;

      if (escalated) {
        console.log(`    [${config.id}] 🔄 Model escalated: ${heuristicModel} → ${model}`);
        console.log(`    [${config.id}] Reason: ${escalation.reason}`);
      }
    }

    console.log(`    [${config.id}] Prompt written (${prompt.length} chars) → ${model}`);

    const result = await executeReviewerSession(
      promptFile,
      {
        cwd: options.cwd,
        timeout: options.timeoutMs,
      },
      config.type,
      model // Pass escalated model
    );

    const parsed = parseReviewResult(config, result, Date.now() - startTime);

    // Add model to result for tracking
    parsed.model = model;

    // Filter findings by minimum severity if configured
    if (config.minSeverity && parsed.findings.length > 0) {
      parsed.findings = filterFindingsBySeverity(parsed.findings, config.minSeverity);
    }

    return parsed;
  } catch (error) {
    return {
      reviewerId: config.id,
      reviewerType: config.type,
      outcome: 'error',
      findings: [],
      summary: `Reviewer error: ${(error as Error).message}`,
      confidence: 0,
      durationMs: Date.now() - startTime,
      error: (error as Error).message,
    };
  } finally {
    try {
      await unlink(promptFile);
    } catch {
      // Ignore cleanup errors
    }
  }
}

interface ReviewerSessionResult {
  exitCode: number;
  output: string;
  error: string | null;
}

/**
 * Execute Claude Code for review.
 */
async function executeReviewerSession(
  promptFile: string,
  options: { cwd: string; timeout: number },
  reviewerType: ReviewerType,
  model: 'opus' | 'sonnet' | 'haiku'
): Promise<ReviewerSessionResult> {
  return new Promise(async (resolve) => {
    const promptContent = await readFile(promptFile, 'utf-8');

    // Don't use --output-format json - we want the raw model output for parsing
    const args = ['-p', '--dangerously-skip-permissions', '--model', model];

    let output = '';
    let errorOutput = '';

    const proc = spawn('claude', args, {
      cwd: options.cwd,
      env: { ...process.env },
      stdio: ['pipe', 'pipe', 'pipe'],
    });

    // Write prompt to stdin and close it
    if (proc.stdin) {
      proc.stdin.write(promptContent);
      proc.stdin.end();
    } else {
      resolve({
        exitCode: -1,
        output: '',
        error: 'Failed to open stdin pipe',
      });
      return;
    }

    const timeoutId = setTimeout(() => {
      proc.kill('SIGTERM');
      resolve({
        exitCode: -1,
        output,
        error: 'Reviewer timed out',
      });
    }, options.timeout);

    proc.stdout?.on('data', (data) => {
      output += data.toString();
    });

    proc.stderr?.on('data', (data) => {
      errorOutput += data.toString();
    });

    proc.on('close', (code) => {
      clearTimeout(timeoutId);
      resolve({
        exitCode: code || 0,
        output,
        error: code !== 0 ? errorOutput || `Exit code ${code}` : null,
      });
    });

    proc.on('error', (err) => {
      clearTimeout(timeoutId);
      resolve({
        exitCode: -1,
        output,
        error: err.message,
      });
    });
  });
}

// ─────────────────────────────────────────────────────────────────────────────
// Result Parsing
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Parse the raw output into a structured ReviewResult.
 * The model should output JSON directly based on the prompt instructions.
 */
function parseReviewResult(
  config: ReviewerConfig,
  result: ReviewerSessionResult,
  durationMs: number
): ReviewResult {
  // Check for session-level errors first
  if (result.error || result.exitCode !== 0) {
    const errorMsg = result.error || `Exit code ${result.exitCode}`;
    // If there's output despite error, try to parse it
    if (!result.output.trim()) {
      return {
        reviewerId: config.id,
        reviewerType: config.type,
        outcome: 'error',
        findings: [],
        summary: `Reviewer session failed: ${errorMsg}`,
        confidence: 0,
        durationMs,
        error: errorMsg,
      };
    }
  }

  // Check for empty output
  if (!result.output.trim()) {
    return {
      reviewerId: config.id,
      reviewerType: config.type,
      outcome: 'error',
      findings: [],
      summary: 'Reviewer produced no output',
      confidence: 0,
      durationMs,
      error: 'Empty output',
    };
  }

  try {
    // Try to extract JSON from the output
    const jsonContent = extractJsonFromOutput(result.output);
    if (!jsonContent) {
      throw new Error('No valid JSON found in reviewer output');
    }

    const parsed = JSON.parse(jsonContent);
    return buildReviewResult(config, parsed, durationMs);
  } catch (error) {
    // If parsing fails, return error result with output snippet for debugging
    const outputPreview = result.output.slice(0, 200).replace(/\n/g, ' ');
    return {
      reviewerId: config.id,
      reviewerType: config.type,
      outcome: 'error',
      findings: [],
      summary: `Failed to parse reviewer output: ${(error as Error).message}`,
      confidence: 0,
      durationMs,
      error: `${(error as Error).message} - Output preview: ${outputPreview}`,
    };
  }
}

/**
 * Extract JSON object from Claude output.
 * Handles code blocks, raw JSON, and mixed content.
 */
function extractJsonFromOutput(output: string): string | null {
  // Strategy 1: Try to find JSON in a code block (```json ... ```)
  const codeBlockMatch = output.match(/```json?\s*([\s\S]*?)\s*```/);
  if (codeBlockMatch) {
    const content = codeBlockMatch[1].trim();
    if (isValidJson(content)) {
      return content;
    }
  }

  // Strategy 2: Find JSON object with balanced braces
  const jsonContent = extractBalancedJson(output);
  if (jsonContent && isValidJson(jsonContent)) {
    return jsonContent;
  }

  // Strategy 3: Try the entire output if it looks like JSON
  const trimmed = output.trim();
  if (trimmed.startsWith('{') && trimmed.endsWith('}') && isValidJson(trimmed)) {
    return trimmed;
  }

  return null;
}

/**
 * Extract a balanced JSON object from text.
 * Finds the first { and matches until balanced }.
 */
function extractBalancedJson(text: string): string | null {
  const startIdx = text.indexOf('{');
  if (startIdx === -1) return null;

  let depth = 0;
  let inString = false;
  let escapeNext = false;

  for (let i = startIdx; i < text.length; i++) {
    const char = text[i];

    if (escapeNext) {
      escapeNext = false;
      continue;
    }

    if (char === '\\' && inString) {
      escapeNext = true;
      continue;
    }

    if (char === '"') {
      inString = !inString;
      continue;
    }

    if (!inString) {
      if (char === '{') {
        depth++;
      } else if (char === '}') {
        depth--;
        if (depth === 0) {
          return text.slice(startIdx, i + 1);
        }
      }
    }
  }

  return null; // Unbalanced braces
}

/**
 * Quick JSON validation check.
 */
function isValidJson(text: string): boolean {
  try {
    JSON.parse(text);
    return true;
  } catch {
    return false;
  }
}

/**
 * Build a ReviewResult from parsed JSON.
 */
function buildReviewResult(
  config: ReviewerConfig,
  parsed: Record<string, unknown>,
  durationMs: number
): ReviewResult {
  // Validate that we have minimum required fields
  if (!parsed || typeof parsed !== 'object') {
    return {
      reviewerId: config.id,
      reviewerType: config.type,
      outcome: 'error',
      findings: [],
      summary: 'Invalid JSON structure: not an object',
      confidence: 0,
      durationMs,
      error: 'Invalid JSON structure',
    };
  }

  const outcome = validateOutcome(parsed.outcome as string);
  const findings = parseFindings(config.id, parsed.findings as unknown[]);

  // Validate and normalize confidence (should be 0-1)
  let confidence: number;
  if (typeof parsed.confidence === 'number') {
    confidence = Math.max(0, Math.min(1, parsed.confidence));
  } else {
    // If no confidence provided, derive from outcome
    confidence = outcome === 'pass' ? 0.9 : outcome === 'pass_with_findings' ? 0.7 : outcome === 'fail' ? 0.6 : 0;
  }

  return {
    reviewerId: config.id,
    reviewerType: config.type,
    outcome,
    findings,
    summary: (parsed.summary as string) || 'No summary provided',
    confidence,
    durationMs,
  };
}

/**
 * Validate and normalize outcome.
 */
function validateOutcome(outcome: string | undefined): ReviewOutcome {
  const valid: ReviewOutcome[] = ['pass', 'pass_with_findings', 'fail', 'error'];
  if (outcome && valid.includes(outcome as ReviewOutcome)) {
    return outcome as ReviewOutcome;
  }
  return 'error';
}

/**
 * Parse findings array from JSON.
 */
function parseFindings(reviewerId: string, findings: unknown[]): ReviewFinding[] {
  if (!Array.isArray(findings)) return [];

  const results: ReviewFinding[] = [];

  for (let i = 0; i < findings.length; i++) {
    const f = findings[i];
    if (typeof f !== 'object' || f === null) continue;

    const finding = f as Record<string, unknown>;

    const result: ReviewFinding = {
      id: `${reviewerId}-${i}`,
      severity: validateSeverity(finding.severity as string),
      category: (finding.category as string) || 'general',
      title: (finding.title as string) || 'Untitled finding',
      description: (finding.description as string) || '',
    };

    // Only add optional properties if they have values
    if (typeof finding.file === 'string') {
      result.file = finding.file;
    }
    if (typeof finding.line === 'number') {
      result.line = finding.line;
    }
    if (typeof finding.suggestion === 'string') {
      result.suggestion = finding.suggestion;
    }

    results.push(result);
  }

  return results;
}

/**
 * Validate and normalize severity.
 */
function validateSeverity(severity: string | undefined): FindingSeverity {
  const valid: FindingSeverity[] = ['critical', 'high', 'medium', 'low', 'info'];
  if (severity && valid.includes(severity as FindingSeverity)) {
    return severity as FindingSeverity;
  }
  return 'info';
}

// ─────────────────────────────────────────────────────────────────────────────
// Utility Functions
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Filter findings by minimum severity.
 */
export function filterFindingsBySeverity(
  findings: ReviewFinding[],
  minSeverity: FindingSeverity
): ReviewFinding[] {
  const severityOrder: FindingSeverity[] = ['critical', 'high', 'medium', 'low', 'info'];
  const minIndex = severityOrder.indexOf(minSeverity);

  return findings.filter((f) => severityOrder.indexOf(f.severity) <= minIndex);
}

/**
 * Get icon for review outcome.
 */
export function getOutcomeIcon(outcome: ReviewOutcome): string {
  switch (outcome) {
    case 'pass':
      return '✅';
    case 'pass_with_findings':
      return '⚠️';
    case 'fail':
      return '❌';
    case 'error':
      return '💥';
  }
}

File: .symphony/workspaces/code-quality/lm-4e95ac36/.symphony/workspaces/code-quality/lm-d9b9f436/packages/agent-sdk/src/create_something_agents/reviewers/reviewer.py
"""
Code Reviewer

Multi-provider code review with specialized reviewers.
Philosophy: Route to the most cost-effective provider that can handle the task.

Provider Routing:
- Security: Gemini Flash (pattern detection) → Haiku fallback
- Quality: Haiku (balanced review) → Sonnet fallback
- Architecture: Sonnet (deep analysis) → Opus fallback
"""

import json
import re
import time
from dataclasses import dataclass
from typing import Any

from ..providers.base import ProviderConfig, ProviderResult
from .types import (
    ReviewerType,
    ReviewOutcome,
    FindingSeverity,
    ReviewFinding,
    ReviewResult,
    ReviewContext,
    ReviewerConfig,
    filter_findings_by_severity,
)
from .prompts import get_prompt_for_reviewer


@dataclass
class ProviderPreference:
    """Provider preference for a reviewer type."""

    primary_provider: str  # "gemini" or "claude"
    primary_model: str
    fallback_provider: str
    fallback_model: str


# Multi-provider routing by reviewer type
REVIEWER_ROUTING: dict[ReviewerType, ProviderPreference] = {
    # Security: Pattern detection is cheap - use Gemini Flash
    ReviewerType.SECURITY: ProviderPreference(
        primary_provider="gemini",
        primary_model="gemini-2.0-flash-exp",
        fallback_provider="claude",
        fallback_model="claude-3-5-haiku-20241022",
    ),
    # Quality: Balanced review - use Haiku, fallback to Sonnet
    ReviewerType.QUALITY: ProviderPreference(
        primary_provider="claude",
        primary_model="claude-3-5-haiku-20241022",
        fallback_provider="claude",
        fallback_model="claude-sonnet-4-20250514",
    ),
    # Architecture: Deep analysis needs reasoning - use Sonnet
    ReviewerType.ARCHITECTURE: ProviderPreference(
        primary_provider="claude",
        primary_model="claude-sonnet-4-20250514",
        fallback_provider="claude",
        fallback_model="claude-opus-4-20250514",
    ),
    # Custom: Default to Sonnet
    ReviewerType.CUSTOM: ProviderPreference(
        primary_provider="claude",
        primary_model="claude-sonnet-4-20250514",
        fallback_provider="claude",
        fallback_model="claude-opus-4-20250514",
    ),
}

# Cost estimates per request (approximate)
COST_ESTIMATES = {
    "gemini-2.0-flash-exp": 0.0001,
    "gemini-1.5-flash": 0.0001,
    "gemini-1.5-pro": 0.0005,
    "claude-3-5-haiku-20241022": 0.001,
    "claude-sonnet-4-20250514": 0.01,
    "claude-opus-4-20250514": 0.10,
}


class CodeReviewer:
    """
    Multi-provider code reviewer.

    Routes reviews to the most cost-effective provider:
    - Security: Gemini Flash (~$0.0001) for pattern detection
    - Quality: Haiku (~$0.001) for balanced review
    - Architecture: Sonnet (~$0.01) for deep analysis
    """

    def __init__(
        self,
        claude_provider: Any | None = None,
        gemini_provider: Any | None = None,
    ):
        """
        Initialize the code reviewer.

        Args:
            claude_provider: ClaudeProvider instance (optional)
            gemini_provider: GeminiProvider instance (optional)
        """
        self.claude_provider = claude_provider
        self.gemini_provider = gemini_provider
        self._failure_counts: dict[str, int] = {}

    def _get_provider(self, provider_name: str) -> Any | None:
        """Get provider by name."""
        if provider_name == "claude":
            return self.claude_provider
        elif provider_name == "gemini":
            return self.gemini_provider
        return None

    def _should_escalate(self, reviewer_id: str, threshold: int = 2) -> bool:
        """Check if we should escalate to fallback due to failures."""
        return self._failure_counts.get(reviewer_id, 0) >= threshold

    def _record_failure(self, reviewer_id: str) -> None:
        """Record a failure for escalation tracking."""
        self._failure_counts[reviewer_id] = self._failure_counts.get(reviewer_id, 0) + 1

    def _clear_failures(self, reviewer_id: str) -> None:
        """Clear failure count on success."""
        self._failure_counts.pop(reviewer_id, None)

    async def review(
        self,
        config: ReviewerConfig,
        context: ReviewContext,
        timeout_ms: int = 120000,
    ) -> ReviewResult:
        """
        Run a code review.

        Args:
            config: Reviewer configuration
            context: Review context (files, diff, etc.)
            timeout_ms: Timeout in milliseconds

        Returns:
            ReviewResult with findings
        """
        start_time = time.time()

        # Generate prompt
        base_prompt = get_prompt_for_reviewer(config.type, config.custom_prompt)
        prompt = self._build_prompt(base_prompt, context)

        # Determine provider and model
        routing = REVIEWER_ROUTING.get(config.type, REVIEWER_ROUTING[ReviewerType.CUSTOM])

        # Check for overrides
        if config.provider_override:
            provider_name = config.provider_override
            model = config.model_override or routing.primary_model
        elif self._should_escalate(config.id):
            provider_name = routing.fallback_provider
            model = config.model_override or routing.fallback_model
            print(f"    [{config.id}] Escalating to {provider_name}/{model}")
        else:
            provider_name = routing.primary_provider
            model = config.model_override or routing.primary_model

        provider = self._get_provider(provider_name)

        # Fallback if primary provider not available
        if provider is None:
            fallback_name = routing.fallback_provider
            provider = self._get_provider(fallback_name)
            if provider is not None:
                provider_name = fallback_name
                model = routing.fallback_model
                print(f"    [{config.id}] Primary provider unavailable, using {provider_name}")

        if provider is None:
            duration_ms = (time.time() - start_time) * 1000
            return ReviewResult(
                reviewer_id=config.id,
                reviewer_type=config.type,
                outcome=ReviewOutcome.ERROR,
                findings=[],
                summary="No provider available",
                confidence=0.0,
                duration_ms=duration_ms,
                error="No provider available (neither Claude nor Gemini configured)",
            )

        # Execute review
        try:
            provider_config = ProviderConfig(
                task=prompt,
                model=model,
                max_tokens=4096,
                temperature=0.0,
            )

            result = await provider.execute(provider_config)
            duration_ms = (time.time() - start_time) * 1000

            if not result.success:
                self._record_failure(config.id)
                return ReviewResult(
                    reviewer_id=config.id,
                    reviewer_type=config.type,
                    outcome=ReviewOutcome.ERROR,
                    findings=[],
                    summary=f"Provider execution failed: {result.error}",
                    confidence=0.0,
                    duration_ms=duration_ms,
                    model=model,
                    provider=provider_name,
                    cost_usd=result.cost_usd,
                    error=result.error,
                )

            # Parse result
            parsed = self._parse_review_output(config, result.output)

            # Build result
            review_result = ReviewResult(
                reviewer_id=config.id,
                reviewer_type=config.type,
                outcome=parsed.get("outcome", ReviewOutcome.ERROR),
                findings=parsed.get("findings", []),
                summary=parsed.get("summary", "No summary provided"),
                confidence=parsed.get("confidence", 0.0),
                duration_ms=duration_ms,
                model=model,
                provider=provider_name,
                cost_usd=result.cost_usd,
            )

            # Filter findings by severity if configured
            if config.min_severity and review_result.findings:
                review_result.findings = filter_findings_by_severity(
                    review_result.findings,
                    config.min_severity,
                )

            self._clear_failures(config.id)
            return review_result

        except Exception as e:
            self._record_failure(config.id)
            duration_ms = (time.time() - start_time) * 1000
            return ReviewResult(
                reviewer_id=config.id,
                reviewer_type=config.type,
                outcome=ReviewOutcome.ERROR,
                findings=[],
                summary=f"Review failed: {str(e)}",
                confidence=0.0,
                duration_ms=duration_ms,
                error=str(e),
            )

    def _build_prompt(self, base_prompt: str, context: ReviewContext) -> str:
        """Build the full prompt with context."""
        context_lines: list[str] = []

        context_lines.append("## Files Changed")
        if context.files_changed:
            for f in context.files_changed:
                context_lines.append(f"- {f}")
        else:
            context_lines.append("(No files changed)")
        context_lines.append("")

        context_lines.append("## Git Diff (This Checkpoint)")
        context_lines.append("```diff")
        # Truncate diff to avoid context overflow
        truncated_diff = context.git_diff[:30000]
        context_lines.append(truncated_diff)
        if len(context.git_diff) > 30000:
            context_lines.append("... (diff truncated)")
        context_lines.append("```")
        context_lines.append("")

        # For architecture, include full harness diff
        if context.full_harness_diff:
            context_lines.append("## Full Harness Diff (All Changes Since Branch Start)")
            context_lines.append(
                "**IMPORTANT: Check this for DRY violations across all files.**"
            )
            context_lines.append("```diff")
            truncated_full = context.full_harness_diff[:20000]
            context_lines.append(truncated_full)
            if len(context.full_harness_diff) > 20000:
                context_lines.append("... (full harness diff truncated)")
            context_lines.append("```")
            context_lines.append("")

        context_lines.append("## Completed Issues")
        if context.completed_issues:
            for issue in context.completed_issues:
                context_lines.append(f"- {issue.get('id', '?')}: {issue.get('title', '?')}")
        else:
            context_lines.append("(No issues completed)")
        context_lines.append("")

        context_lines.append("## Checkpoint Summary")
        context_lines.append(context.checkpoint_summary or "(No summary available)")

        context_section = "\n".join(context_lines)
        return base_prompt.replace("{CONTEXT}", context_section)

    def _parse_review_output(
        self,
        config: ReviewerConfig,
        output: str,
    ) -> dict[str, Any]:
        """Parse reviewer output into structured data."""
        # Extract JSON from output
        json_content = self._extract_json(output)
        if not json_content:
            return {
                "outcome": ReviewOutcome.ERROR,
                "findings": [],
                "summary": "Failed to parse reviewer output (no JSON found)",
                "confidence": 0.0,
            }

        try:
            parsed = json.loads(json_content)
        except json.JSONDecodeError as e:
            return {
                "outcome": ReviewOutcome.ERROR,
                "findings": [],
                "summary": f"Invalid JSON: {e}",
                "confidence": 0.0,
            }

        # Validate outcome
        outcome_str = parsed.get("outcome", "error")
        outcome_map = {
            "pass": ReviewOutcome.PASS,
            "pass_with_findings": ReviewOutcome.PASS_WITH_FINDINGS,
            "fail": ReviewOutcome.FAIL,
            "error": ReviewOutcome.ERROR,
        }
        outcome = outcome_map.get(outcome_str, ReviewOutcome.ERROR)

        # Parse findings
        findings = self._parse_findings(config.id, parsed.get("findings", []))

        # Validate confidence
        confidence = parsed.get("confidence", 0.0)
        if not isinstance(confidence, (int, float)):
            confidence = 0.5
        confidence = max(0.0, min(1.0, float(confidence)))

        return {
            "outcome": outcome,
            "findings": findings,
            "summary": parsed.get("summary", "No summary provided"),
            "confidence": confidence,
        }

    def _extract_json(self, output: str) -> str | None:
        """Extract JSON from model output."""
        output = output.strip()

        # Handle prefilled responses (output may not start with {)
        if not output.startswith("{"):
            output = '{\n  "outcome":' + output

        # Strategy 1: Code block
        code_block = re.search(r"```json?\s*([\s\S]*?)\s*```", output)
        if code_block:
            content = code_block.group(1).strip()
            if self._is_valid_json(content):
                return content

        # Strategy 2: Balanced braces
        json_content = self._extract_balanced_json(output)
        if json_content and self._is_valid_json(json_content):
            return json_content

        # Strategy 3: Whole output
        if output.startswith("{") and output.endswith("}"):
            if self._is_valid_json(output):
                return output

        return None

    def _extract_balanced_json(self, text: str) -> str | None:
        """Extract balanced JSON object from text."""
        start_idx = text.find("{")
        if start_idx == -1:
            return None

        depth = 0
        in_string = False
        escape_next = False

        for i in range(start_idx, len(text)):
            char = text[i]

            if escape_next:
                escape_next = False
                continue

            if char == "\\" and in_string:
                escape_next = True
                continue

            if char == '"':
                in_string = not in_string
                continue

            if not in_string:
                if char == "{":
                    depth += 1
                elif char == "}":
                    depth -= 1
                    if depth == 0:
                        return text[start_idx : i + 1]

        return None

    def _is_valid_json(self, text: str) -> bool:
        """Check if text is valid JSON."""
        try:
            json.loads(text)
            return True
        except json.JSONDecodeError:
            return False

    def _parse_findings(
        self,
        reviewer_id: str,
        findings: list[Any],
    ) -> list[ReviewFinding]:
        """Parse findings array from JSON."""
        if not isinstance(findings, list):
            return []

        results: list[ReviewFinding] = []

        for i, f in enumerate(findings):
            if not isinstance(f, dict):
                continue

            # Validate severity
            severity_str = f.get("severity", "info")
            severity_map = {
                "critical": FindingSeverity.CRITICAL,
                "high": FindingSeverity.HIGH,
                "medium": FindingSeverity.MEDIUM,
                "low": FindingSeverity.LOW,
                "info": FindingSeverity.INFO,
            }
            severity = severity_map.get(severity_str, FindingSeverity.INFO)

            finding = ReviewFinding(
                id=f"{reviewer_id}-{i}",
                severity=severity,
                category=f.get("category", "general"),
                title=f.get("title", "Untitled finding"),
                description=f.get("description", ""),
                file=f.get("file"),
                line=f.get("line"),
                quote=f.get("quote"),
                suggestion=f.get("suggestion"),
            )
            results.append(finding)

        return results


async def run_review(
    config: ReviewerConfig,
    context: ReviewContext,
    claude_provider: Any | None = None,
    gemini_provider: Any | None = None,
    timeout_ms: int = 120000,
) -> ReviewResult:
    """
    Convenience function to run a single review.

    Args:
        config: Reviewer configuration
        context: Review context
        claude_provider: Optional ClaudeProvider
        gemini_provider: Optional GeminiProvider
        timeout_ms: Timeout in milliseconds

    Returns:
        ReviewResult
    """
    reviewer = CodeReviewer(claude_provider, gemini_provider)
    return await reviewer.review(config, context, timeout_ms)
